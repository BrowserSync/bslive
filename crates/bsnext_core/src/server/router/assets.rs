use crate::server::state::ServerState;
use axum::response::IntoResponse;
use axum::routing::get;
use axum::Router;
use bsnext_client::{UI_CSS, UI_JS};
use http::header::CONTENT_TYPE;
use http::HeaderValue;
use mime_guess::mime;
use std::sync::Arc;

pub fn pub_ui_assets(_: Arc<ServerState>) -> Router<Arc<ServerState>> {
    Router::new()
        .route(
            "/dist/index.js",
            get(|| async {
                (
                    [(
                        CONTENT_TYPE,
                        HeaderValue::from_static(mime::APPLICATION_JAVASCRIPT.as_ref()),
                    )],
                    UI_JS,
                )
                    .into_response()
            }),
        )
        .route(
            "/dist/index.css",
            get(|| async {
                (
                    [(
                        CONTENT_TYPE,
                        HeaderValue::from_static(mime::TEXT_CSS.as_ref()),
                    )],
                    UI_CSS,
                )
                    .into_response()
            }),
        )
}
