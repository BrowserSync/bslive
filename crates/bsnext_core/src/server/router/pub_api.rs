use crate::dto::{RouteDTO, ServerDesc};
use crate::server::state::ServerState;
use axum::extract::State;
use axum::response::IntoResponse;
use axum::routing::get;
use axum::{Json, Router};
use http::Uri;
use std::sync::Arc;

async fn api_handler(State(app): State<Arc<ServerState>>, _uri: Uri) -> impl IntoResponse {
    let routes = app.routes.read().await;
    let dto = ServerDesc {
        routes: routes.iter().map(|r| RouteDTO::from(r)).collect(),
    };
    Json(dto)
}

pub fn pub_api(state: Arc<ServerState>) -> Router<Arc<ServerState>> {
    Router::new()
        .route("/servers", get(api_handler))
        .with_state(state.clone())
}
