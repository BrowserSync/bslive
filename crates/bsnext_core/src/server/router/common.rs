use crate::server::router::make_router;
use crate::server::state::ServerState;
use std::net::SocketAddr;

use axum::body::Body;
use axum::extract::Request;
use axum::response::Response;
use axum::Router;
use bsnext_dto::ClientEvent;
use bsnext_input::server_config::ServerConfig;
use bsnext_input::Input;
use http::header::ACCEPT;
use http::response::Parts;
use http::HeaderValue;
use mime_guess::mime::TEXT_HTML;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;
use tokio::task::JoinHandle;
use tower::ServiceExt;

pub fn into_state(val: ServerConfig) -> ServerState {
    let (sender, _) = tokio::sync::broadcast::channel::<ClientEvent>(10);
    ServerState {
        routes: Arc::new(RwLock::new(val.routes.clone())),
        raw_router: Arc::new(RwLock::new(Router::new())),
        id: val.identity.as_id(),
        parent: None,
        evt_receiver: None,
        client_sender: Arc::new(sender),
    }
}

pub async fn to_resp_body(res: Response) -> String {
    use http_body_util::BodyExt;
    let (_parts, body) = res.into_parts();
    let b = body.collect().await.unwrap();
    let b = b.to_bytes();
    let as_str = std::str::from_utf8(&b).unwrap();
    as_str.to_owned()
}

pub async fn to_resp_parts_and_body(res: Response) -> (Parts, String) {
    use http_body_util::BodyExt;
    let (parts, body) = res.into_parts();
    let b = body.collect().await.unwrap();
    let b = b.to_bytes();
    let as_str = std::str::from_utf8(&b).unwrap();
    (parts, as_str.to_owned())
}

pub async fn req_to_body(state: ServerState, uri: &str) -> String {
    let app = make_router(&Arc::new(state));
    let req = Request::get(uri).body(Body::empty()).unwrap();
    let res = app.oneshot(req).await.unwrap();
    to_resp_body(res).await
}

pub async fn accept_html_req_to_body(state: ServerState, uri: &str) -> String {
    let app = make_router(&Arc::new(state));
    let mut req = Request::get(uri).body(Body::empty()).unwrap();
    req.headers_mut()
        .insert(ACCEPT, HeaderValue::from_str(TEXT_HTML.as_ref()).unwrap());
    let res = app.oneshot(req).await.unwrap();
    to_resp_body(res).await
}

pub async fn uri_to_res(state: ServerState, uri: &str) -> Response {
    let app = make_router(&Arc::new(state));
    let req = Request::get(uri).body(Body::empty()).unwrap();
    app.oneshot(req).await.unwrap()
}

pub async fn uri_to_res_parts(state: ServerState, uri: &str) -> (Parts, String, Duration) {
    let app = make_router(&Arc::new(state));
    let req = Request::get(uri).body(Body::empty()).unwrap();
    let start = Instant::now();
    let res = app.oneshot(req).await.unwrap();
    let end = Instant::now();
    let diff = end - start;
    let (parts, body) = to_resp_parts_and_body(res).await;
    (parts, body, diff)
}

pub fn from_yaml(yaml: &str) -> anyhow::Result<ServerState> {
    let input: Input = serde_yaml::from_str(yaml)?;
    let config: ServerConfig = input.servers.first().expect("first").to_owned();
    let state = into_state(config);
    Ok(state)
}

pub struct TestProxy {
    pub socker_addr: SocketAddr,
    pub http_addr: String,
    pub shutdown: tokio::sync::oneshot::Sender<()>,
    pub join_handle: JoinHandle<()>,
}

impl TestProxy {
    pub async fn destroy(self) -> anyhow::Result<()> {
        self.shutdown.send(()).expect("did send");
        self.join_handle.await?;
        Ok(())
    }
}

pub async fn test_proxy(router: Router) -> anyhow::Result<TestProxy> {
    let (address_sender, address_receiver) = tokio::sync::oneshot::channel::<SocketAddr>();
    let (complete_sender, complete_receiver) = tokio::sync::oneshot::channel::<()>();

    let handle = tokio::spawn(async move {
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let socket_addr = listener.local_addr().unwrap();

        // give consumers the address
        address_sender.send(socket_addr).expect("can send");

        // serve and wait for shutdown
        axum::serve(listener, router)
            .with_graceful_shutdown(async {
                complete_receiver.await.expect("TODO: panic message");
            })
            .await
            .unwrap();
    });

    let addr = address_receiver.await?;
    let http_address = format!("http://{addr}");
    Ok(TestProxy {
        socker_addr: addr,
        http_addr: http_address,
        shutdown: complete_sender,
        join_handle: handle,
    })
}
