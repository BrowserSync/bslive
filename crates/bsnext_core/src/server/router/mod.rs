pub mod common;

use crate::panic_handler::handle_panic;
use axum::extract::{Request, State};
use axum::http::Uri;
use axum::middleware::{from_fn_with_state, Next};
use axum::response::{IntoResponse, Response};
use axum::routing::{get, MethodRouter};
use axum::{http, middleware, Extension, Router};

use axum::body::Body;
use http::header::CONTENT_TYPE;
use http::{HeaderValue, StatusCode};
use hyper_tls::HttpsConnector;
use hyper_util::client::legacy::connect::HttpConnector;
use hyper_util::client::legacy::Client;
use hyper_util::rt::TokioExecutor;
use mime_guess::mime;
use std::sync::Arc;
use tower::{ServiceBuilder, ServiceExt};
use tower_http::catch_panic::CatchPanicLayer;

use crate::meta::MetaData;
use crate::server::router::assets::pub_ui_assets;
use crate::server::router::pub_api::pub_api;
use crate::server::state::ServerState;
use crate::ws::ws_handler;
use bsnext_client::html_with_base;
use bsnext_dto::{RouteDTO, ServerDesc};
use tracing::{span, Level};

mod assets;
mod pub_api;

pub fn make_router(state: &Arc<ServerState>) -> Router {
    let https = HttpsConnector::new();
    let client: Client<HttpsConnector<HttpConnector>, Body> =
        Client::builder(TokioExecutor::new()).build(https);

    let router = Router::new()
        .merge(built_ins(state.clone()))
        .merge(dynamic_loaders(state.clone()));

    router
        // todo(alpha): enable tracing on a per-item basis?
        // .layer(TraceLayer::new_for_http())
        // todo(alpha): re-enable in the correct place?
        // .layer(TimeoutLayer::new(Duration::from_secs(10)))
        .layer(Extension(client))
    // todo: When to add this compression back in?
    // .layer(CompressionLayer::new())
}

pub fn built_ins(state: Arc<ServerState>) -> Router {
    async fn handler(State(app): State<Arc<ServerState>>, _uri: Uri) -> impl IntoResponse {
        // let v = app.lock().await;
        let routes = app.routes.read().await;
        let _dto = ServerDesc {
            routes: routes.iter().map(RouteDTO::from).collect(),
            id: app.id.to_string(),
        };
        let markup = html_with_base("/__bs_assets/ui/");
        (
            [(
                CONTENT_TYPE,
                HeaderValue::from_static(mime::TEXT_HTML_UTF_8.as_ref()),
            )],
            markup,
        )
            .into_response()
    }
    async fn js_handler(_uri: Uri) -> impl IntoResponse {
        let markup = include_str!("../../../../../inject/dist/index.js");
        (
            [(
                CONTENT_TYPE,
                HeaderValue::from_static(mime::APPLICATION_JAVASCRIPT_UTF_8.as_ref()),
            )],
            markup,
        )
            .into_response()
    }

    route("/__bslive", get(handler))
        .route("/__bs_js", get(js_handler))
        .route("/__bs_ws", get(ws_handler))
        .nest("/__bs_api", pub_api(state.clone()))
        .nest("/__bs_assets/ui", pub_ui_assets(state.clone()))
        .with_state(state.clone())
}

fn route(path: &str, method_router: MethodRouter<Arc<ServerState>>) -> Router<Arc<ServerState>> {
    Router::new().route(path, method_router)
}

pub fn dynamic_loaders(state: Arc<ServerState>) -> Router {
    Router::new()
        .layer(
            ServiceBuilder::new()
                .layer(from_fn_with_state(state.clone(), tagging_layer))
                // .layer(from_fn_with_state(state.clone(), maybe_proxy))
                // todo(alpha): have the order listed here instead: static -> dir -> proxy
                // .layer(from_fn_with_state(state.clone(), not_found_loader))
                .layer(from_fn_with_state(state.clone(), dynamic_router)),
        )
        .layer(CatchPanicLayer::custom(handle_panic))
        .with_state(state.clone())
}

async fn tagging_layer(
    req: Request,
    next: Next,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let mut r = next.run(req).await;
    #[allow(unused_variables)]
    if let Some(md) = r.extensions_mut().get::<MetaData>() {
        // r.headers_mut()
        //     .insert("abc", HeaderValue::from_static(md.as_header_value()));
    }
    Ok(r)
}

#[derive(serde::Deserialize, Debug)]
pub enum Encoding {
    Br,
    Zip,
}

pub async fn dynamic_router(
    State(app): State<Arc<ServerState>>,
    req: Request,
    _next: Next,
) -> impl IntoResponse {
    let span = span!(parent: None, Level::INFO, "raw_loader", path = req.uri().path());
    let _guard = span.enter();

    let raw_router = app.raw_router.read().await;

    raw_router
        .clone()
        .layer(middleware::from_fn(tag_dynamic))
        .oneshot(req)
        .await
        .into_response()
}

async fn tag_dynamic(req: Request, next: Next) -> Result<impl IntoResponse, (StatusCode, String)> {
    let (mut parts, body) = next.run(req).await.into_parts();
    if parts.status.as_u16() == 200 {
        parts.extensions.insert(MetaData::ServedRaw);
    }
    Ok(Response::from_parts(parts, body))
}
