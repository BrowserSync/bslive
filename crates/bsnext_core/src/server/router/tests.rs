#[cfg(test)]
pub mod common {
    use crate::server::router::make_router;
    use crate::server::state::ServerState;
    use axum::body::Body;
    use axum::extract::Request;
    use axum::response::Response;
    use bsnext_dto::ClientEvent;
    use bsnext_input::server_config::ServerConfig;
    use http::header::ACCEPT;
    use http::HeaderValue;
    use mime_guess::mime::TEXT_HTML;
    use std::sync::Arc;
    use tokio::sync::RwLock;
    use tower::ServiceExt;

    impl From<ServerConfig> for ServerState {
        fn from(val: ServerConfig) -> ServerState {
            let (sender, _) = tokio::sync::broadcast::channel::<ClientEvent>(10);
            ServerState {
                routes: Arc::new(RwLock::new(val.routes.clone())),
                id: val.identity.as_id(),
                parent: None,
                client_sender: Arc::new(sender),
            }
        }
    }

    pub async fn to_resp_body(res: Response) -> String {
        use http_body_util::BodyExt;
        let (_parts, body) = res.into_parts();
        let b = body.collect().await.unwrap();
        let b = b.to_bytes();
        let as_str = std::str::from_utf8(&b).unwrap();
        as_str.to_owned()
    }

    pub async fn req_to_body(state: ServerState, uri: &str) -> String {
        let app = make_router(&Arc::new(state));
        let req = Request::get(uri).body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();
        let body = to_resp_body(res).await;
        body
    }

    pub async fn accept_html_req_to_body(state: ServerState, uri: &str) -> String {
        let app = make_router(&Arc::new(state));
        let mut req = Request::get(uri).body(Body::empty()).unwrap();
        req.headers_mut().insert(
            ACCEPT,
            HeaderValue::from_str(&TEXT_HTML.to_string()).unwrap(),
        );
        let res = app.oneshot(req).await.unwrap();
        let body = to_resp_body(res).await;
        body
    }
}

#[cfg(test)]
mod test {
    use crate::server::router::make_router;
    use crate::server::router::tests::common::to_resp_body;
    use crate::server::state::ServerState;
    use axum::body::Body;
    use axum::extract::Request;
    
    
    use bsnext_input::route::{CorsOpts, Route, RouteKind};
    use bsnext_input::server_config::{Identity, ServerConfig};
    
    
    
    
    use http::HeaderValue;
    use std::collections::BTreeMap;
    use std::sync::Arc;
    
    use tower::ServiceExt;

    #[tokio::test]
    async fn test_handlers() -> Result<(), anyhow::Error> {
        let headers: BTreeMap<String, String> =
            [("a".to_string(), "b".to_string())].into_iter().collect();
        let state: ServerState = ServerConfig {
            identity: Identity::Address {
                bind_address: "127.0.0.1:3000".to_string(),
            },
            routes: vec![Route {
                path: "/hello".to_string(),
                kind: RouteKind::html("🐥"),
                headers: Some(headers),
                ..Default::default()
            }],
            ..Default::default()
        }
        .into();

        let app = make_router(&Arc::new(state));
        let req = Request::get("/hello").body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();

        insta::assert_debug_snapshot!(res.headers());
        assert_eq!(
            res.headers().get("content-type").unwrap(),
            "text/html; charset=utf-8"
        );

        let body = to_resp_body(res).await;
        assert_eq!(body, "🐥");
        Ok(())
    }

    #[tokio::test]
    async fn test_handlers_raw() -> Result<(), anyhow::Error> {
        let state: ServerState = ServerConfig {
            identity: Identity::Address {
                bind_address: "127.0.0.1".to_string(),
            },
            routes: vec![Route {
                path: "/styles.css".to_string(),
                kind: RouteKind::Raw {
                    raw: "body{}".into(),
                },
                ..Default::default()
            }],
            ..Default::default()
        }
        .into();

        let app = make_router(&Arc::new(state));
        let req = Request::get("/styles.css").body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();

        assert_eq!(res.headers().get("content-length").unwrap(), "6");
        assert_eq!(res.headers().get("content-type").unwrap(), "text/css");

        let body = to_resp_body(res).await;
        assert_eq!(body, "body{}");
        Ok(())
    }

    #[tokio::test]
    async fn test_cors_handlers() -> Result<(), anyhow::Error> {
        let state: ServerState = ServerConfig {
            identity: Identity::Address {
                bind_address: "127.0.0.1:3000".to_string(),
            },
            routes: vec![
                Route {
                    path: "/".to_string(),
                    cors_opts: Some(CorsOpts::Cors(true)),
                    kind: RouteKind::html("home"),
                    ..Default::default()
                },
                Route {
                    path: "/hello".to_string(),
                    kind: RouteKind::html("🐥"),
                    ..Default::default()
                },
            ],
            ..Default::default()
        }
        .into();

        let app = make_router(&Arc::new(state));
        let req = Request::get("/").body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();
        let h = res.headers().get("access-control-allow-origin");
        let v = res.headers().get("vary");

        assert_eq!(h, Some(HeaderValue::from_str("*").as_ref().unwrap()));
        assert_eq!(
            v,
            Some(
                HeaderValue::from_str(
                    "origin, access-control-request-method, access-control-request-headers"
                )
                .as_ref()
                .unwrap()
            )
        );

        let body = to_resp_body(res).await;
        assert_eq!(body, "home");
        Ok(())
    }
    #[tokio::test]
    async fn test_not_found_handler() -> Result<(), anyhow::Error> {
        let state: ServerState = ServerConfig {
            identity: Identity::Address {
                bind_address: "127.0.0.1:3000".to_string(),
            },
            routes: vec![Route {
                path: "/".to_string(),
                cors_opts: Some(CorsOpts::Cors(true)),
                kind: RouteKind::html("home"),
                ..Default::default()
            }],
            ..Default::default()
        }
        .into();

        let app = make_router(&Arc::new(state));
        let req = Request::get("/abc").body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();
        let status = res.status().as_u16();

        assert_eq!(
            res.headers().get("content-type").unwrap(),
            "text/html; charset=utf-8"
        );

        let body = to_resp_body(res).await;

        assert!(body.contains("<title>Browsersync LIVE</title>"));
        assert_eq!(status, 404);
        Ok(())
    }
    #[tokio::test]
    async fn test_route_list() -> Result<(), anyhow::Error> {
        let state: ServerState = ServerConfig {
            identity: Identity::Address {
                bind_address: "127.0.0.1:3000".to_string(),
            },
            routes: vec![Route {
                path: "/abc".to_string(),
                cors_opts: Some(CorsOpts::Cors(true)),
                kind: RouteKind::html("home"),
                ..Default::default()
            }],
            watchers: vec![],
        }
        .into();

        let app = make_router(&Arc::new(state));
        let req = Request::get("/__bslive").body(Body::empty()).unwrap();
        let res = app.oneshot(req).await.unwrap();
        let status = res.status().as_u16();

        assert_eq!(
            res.headers().get("content-type").unwrap(),
            "text/html; charset=utf-8"
        );

        let body = to_resp_body(res).await;

        assert!(body.contains("<title>Browsersync LIVE</title>"));
        assert!(body.contains("<base href=\"/__bs_assets/ui/\" />"));
        assert_eq!(status, 200);
        Ok(())
    }
}
