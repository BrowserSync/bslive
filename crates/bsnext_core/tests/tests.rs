use bsnext_core::server::router::common::{from_yaml, uri_to_res_parts};
use http::header::{CACHE_CONTROL, CONTENT_TYPE};
use http::HeaderValue;

#[tokio::test]
async fn test_handlers() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /hello
            html: "🐥"
            headers:
                a: b
    "#;
    let state = from_yaml(input)?;
    let (parts, body, ..) = uri_to_res_parts(state, "/hello").await;

    insta::assert_debug_snapshot!(parts.headers);

    assert_eq!(body, "🐥");
    Ok(())
}

#[tokio::test]
async fn test_handlers_raw() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /styles.css
            raw: "body{}"
    "#;
    let state = from_yaml(input)?;
    let (parts, body, ..) = uri_to_res_parts(state, "/styles.css").await;

    assert_eq!(parts.headers.get("content-length").unwrap(), "6");
    assert_eq!(parts.headers.get("content-type").unwrap(), "text/css");
    assert_eq!(body, "body{}");

    Ok(())
}

#[tokio::test]
async fn test_cors_handlers() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /
            html: home
            cors: true
    "#;
    let state = from_yaml(input)?;
    let (parts, body, ..) = uri_to_res_parts(state, "/").await;
    let h = parts.headers.get("access-control-allow-origin");
    let v = parts.headers.get("vary");

    assert_eq!(h, Some(HeaderValue::from_str("*").as_ref().unwrap()));
    assert_eq!(
        v,
        Some(
            HeaderValue::from_str(
                "origin, access-control-request-method, access-control-request-headers"
            )
            .as_ref()
            .unwrap()
        )
    );

    assert_eq!(body, "home");
    Ok(())
}
#[tokio::test]
async fn test_route_list() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /abc
            html: home
    "#;
    let state = from_yaml(input)?;
    let (parts, body, ..) = uri_to_res_parts(state, "/__bslive").await;

    let status = parts.status.as_u16();

    assert_eq!(
        parts.headers.get("content-type").unwrap(),
        "text/html; charset=utf-8"
    );

    assert!(body.contains("<title>Browsersync LIVE</title>"));
    assert!(body.contains("<base href=\"/__bs_assets/ui/\" />"));
    assert_eq!(status, 200);
    Ok(())
}

#[tokio::test]
async fn test_cache_prevent() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /
            html: home
    "#;
    let state = from_yaml(input)?;
    let (parts, ..) = uri_to_res_parts(state, "/").await;
    assert_eq!(
        parts.headers.get(CONTENT_TYPE).unwrap(),
        "text/html; charset=utf-8"
    );
    assert_eq!(
        parts.headers.get(CACHE_CONTROL).unwrap(),
        "no-store, no-cache, must-revalidate"
    );
    Ok(())
}

#[tokio::test]
async fn test_cache_default() -> Result<(), anyhow::Error> {
    let input = r#"
    servers:
      - bind_address: 0.0.0.0:3000
        routes:
          - path: /
            html: home
            cache: 'default'
    "#;
    let state = from_yaml(input)?;
    let (parts, ..) = uri_to_res_parts(state, "/").await;
    assert_eq!(parts.headers.get(CACHE_CONTROL), None);
    Ok(())
}
