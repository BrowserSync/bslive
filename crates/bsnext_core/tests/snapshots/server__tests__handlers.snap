---
source: crates/bsnext_core/tests/tests.rs
expression: parts.headers
---
{
    "content-type": "text/html; charset=utf-8",
    "content-length": "4",
    "a": "b",
    "cache-control": "no-store, no-cache, must-revalidate",
    "pragma": "no-cache",
    "expires": "0",
}
